### ISSUE TITLE
Support for counters in esc scripts?

### ISSUE THREAD

**Issue by mjtorn** [godotengine/escoria#66]
It would be useful to have counters in esc scripts. The feature I'm thinking of is when the player tries to do something many times, eg. open a locked door, after a couple of tries he could say "Maybe there's a key somewhere."

I'm not sure, but I think Thimbleweed Park had this kind of a mechanic.

The implications for code could be pretty huge and I'm not sure how to go about implementing it, but here are some ideas.

`incr_global tried_door` would bump the value by one, starting with 1 if `tried_door` is unset.

`decr_global tried_door` would decrement the value. What do you feel about going to 0 or below?

`gt tried_door 3` style RPN notation would evaluate. `gt` for `greater than`. `eq` for `equals` (in case the door is kicked by `use` and takes three times to break in). Characters used in case there are problems using symbols like `>=` which may even confuse the parser into thinking there's a group coming.

The syntax could be like this
```
:use
say player "It's locked."
incr_global tried_door
say player "Maybe there's a key somewhere." [ge tried_door 3]
```

or

```
:use
> [eq kicked_door 3, !door_opened]
    say player "Finally!"
    anim door player_kicks_door_open
    set_global door_opened

> [!door_opened]
    say player "It's stuck."
    anim player player_kicks_door
    incr_global kicked_door
```

Writing these examples also served as an exercise for me, because I haven't done animations and am not sure if I got the syntax in general correct, so please let me know ;)

But the point is that counters have their uses - what do you think?

**Comment by mjtorn** [comment]
Some other things came to mind.

Can inventory items be incremented, like finding up to 8 pieces of eight?

Should the counters be tied to a character instead of being global, in case Escoria starts supporting multiple player characters?

This could potentially cause interesting interactions where an NPC has 8 pieces of eight and the player does things to get all of them.

**Comment by fleskesvor** [comment]
I think this is a good idea, and it's functionality we also talked about for The Farrier. For that game, we ended up just using multiple flags for now. One thing to keep in mind is that the save system currently only supports booleans, so that will also need to be improved, because otherwise you'd lose the current state when you quit the game.

With this functionality in place, incrementing inventory items should be trivial, since they can use an AnimatedSprite and have their state changed when you perform the action which increases the counter.

**Comment by StraToN** [comment]
This is a very good idea. SCUMM language does manage counter like this : http://www.pagetable.com/?p=614

**Comment by lavaduderDev** [comment]
I can see this being great for characters who have multiple comments on something. Like if your annoying some one long enough they just explode into fiery wraith of anger.

**Comment by fleskesvor** [comment]
To avoid making the language too verbose, we could change the signature of `set_global` to take a string as the second parameter, and check if the string looks like a boolean (`true`/`false`), an integer or an arithmetic expression, eg. `+1`, `-2`, `*3` or `/4`.  If the argument is not recognised, we call `report_errors()` and ignore it.

**Comment by fleskesvor** [comment]
I think these use cases should all be solvable once #102 is merged:

> `incr_global tried_door` would bump the value by one, starting with 1 if `tried_door` is unset.

Unset globals default to 0, so `inc_global tried_door 1` would set `tried_door` to 1.

> `decr_global tried_door` would decrement the value. What do you feel about going to 0 or below?

Safe-guarding against decrementing to negative values can be achieved with `dec_global tried_door 1 [gt tried_door 0]`

> `gt tried_door 3` style RPN notation would evaluate. `gt` for `greater than`. `eq` for `equals` (in case the door is kicked by `use` and takes three times to break in). Characters used in case there are problems using symbols like >= which may even confuse the parser into thinking there's a group coming.

`eq`, `gt` and `lt` are implemented. `ne`, `ge` and `le` are not available as operators, but can be used by negating `eq`, `lt` and `gt` respectively.

As it is, `inc_global` and `dec_global` always take an integer value as the second parameter, but maybe they should default to incrementing and decrementing by `1` if a second parameter is not present? The drawback is that we can't use the current infrastructure to force the type of the second parameter, but I think it's OK to assume an integer and break if not, since the function is documented and straight-forward to use.

**Comment by fleskesvor** [comment]
Closed by #102. See updated esc reference for usage.
